### PR TITLE
Releasing the messages immediately if prefetch is disabled and if there is no (active) downstream

### PR DESCRIPTION
**Pull request by anuchandy**
# Description

- Based on the implementation thoughts from Connie on this pr https://github.com/Azure/azure-sdk-for-java/pull/26894
- Releasing the messages immediately if there is no active downstream, which reduces the chances of serving expired messages, see #26632 

## What is not in the scope of this pr
- There might be corner cases where we benefit from doing the release (such as when the client closes/terminates), but those are not in the scope of this pr. This pr explicitly targets the case when there is no active downstream.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by conniey — APPROVED** [review]
Thanks for fixing my PR @anuchandy ! One test case I had troubles testing was whether it released messages if they expired while the same current work was polling and that the next work item would get new messages...

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java [hunk 1]
```diff
@@ -147,43 +172,59 @@ private void drainQueue() {
...
+                    } else {
+                        // Re-buffer the message as it couldn't be emitted or release was disabled.
+                        bufferMessages.addFirst(message);
+                        break;
+                    }
                 }
 
-                numberEmitted++;
+                // account for consumed message - message is either emitted or released.
+                numberConsumed++;
```
Comment (ramya-rao-a) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java:221]: If prefetch is enabled and the message was not emitted, we should not increment this right?

Comment (anuchandy) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java:221]: Exactly, so if the prefetch is enabled and drain couldn't emit downstream, we break from the while-loop in line 216; that way, we don't consider the message as consumed.

**Comment by anuchandy** [comment]
Hi @conniey, I've added a UT. Regarding the expired messages, I later learned from Ramya that we could not release if it's already expired, so release should happen asap if there are no active receivers.

**Comment by anuchandy** [comment]
The below diagram shows what is with this new release feature, adding it for future reference -

Screen Shot 2022-02-10 at 5 16 13 PM

quite_period: a period in the sliding window with no active receivers, any messages during this duration are released.

As illustrated, we continue to use the sliding window, i.e., the prefetch queue is shared, and there is no correlation such as _this_ set of messages belonging to _this_ specific receive call. By design, a future receive call can receive messages that arrived as a result of an earlier credit call (e.g., receive_3); those messages are unlikely to be expired as they just arrived

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java [hunk 2]
```diff
@@ -212,6 +213,8 @@ public static DeliveryState getDeliveryState(DispositionStatus dispositionStatus
                     state = deferredOutcome;
                 }
                 break;
+            case RELEASED:
+                return Released.getInstance();
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:217]:
```suggestion
                state = Released.getInstance();
                break;
```

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java [hunk 3]
```diff
@@ -147,43 +172,59 @@ private void drainQueue() {
...
-                // Push this back to the head of the work queue.
                 if (!isEmitted) {
-                    bufferMessages.addFirst(message);
-                    break;
+                    // The only reason we can't emit was the downstream(s) were terminated hence nobody
+                    // to receive the message.
+                    if (isPrefetchDisabled) {
+                        // release is enabled only for no-prefetch scenario.
+                        asyncClient.release(message).subscribe(__ -> { },
+                            error -> logger.warning("lockToken[{}] Couldn't release the message.",
```
Comment (weidongxu-microsoft) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java:209]: Do we recommend using `logger.atWarning().addKeyValue("lockToken", <>)`?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java (hunk 2 above)
Comment (anuchandy) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java:217]: nice, thanks!

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java (hunk 3 above)
Comment (anuchandy) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java:209]: I think lock-token is something eligible to define as an attribute to lookup, but we haven't migrated the SB package to structured logging yet, so I think we should decide the standard attributes (to use as key-value pairs) when making that effort in the SB package as a whole.

**Pull request merged**